experiments/2018-05-23-rust-on-android-JNA.md

```diff
@@ -0,0 +1,198 @@
+
+```
+cargo build --target aarch64-linux-android --release
+cargo build --target armv7-linux-androideabi --release
+cargo build --target i686-linux-android --release
+```
+
+Now let's create our Android project.
+
+Open Android Studio and select `Start a New Android Project` from the options.
+
+![Create new Android project](assets/2017-09-21-rust-on-android/new-project.png)
+
+On the next screen, type a project name of `Greetings` into the `Application name` field, choose your `Company domain` and select the `android` directory we created earlier as the `Project location`. This will create your Android project inside `greetings/android/Greetings`. Click `Next`.
+
+![Name project](assets/2017-09-21-rust-on-android/name-project.png)
+
+On the next screen, make sure the `Phone and Tablet` option is selected. Click `Next`.
+
+![Choose form factor](assets/2017-09-21-rust-on-android/form-factors.png)
+
+Now we will be asked to choose a starting activity. Select the `Empty Activity` option and click `Next`.
+
+![Choose starting activity](assets/2017-09-21-rust-on-android/add-activity.png)
+
+Name your Activity and layout on the following screen, calling the activity `GreetingsActivity` and the layout `activity_greetings`. Click `Finish`.
+
+![Name activity](assets/2017-09-21-rust-on-android/customize-activity.png)
+
+We now have a build Rust library and an Android project. We now need to link the two. First, we have to add JNA to our project. Open your applications `build.gradle` file and add the following to the `dependencies` block. When prompted, resync gradle.
+
+```
+implementation 'net.java.dev.jna:jna:4.5.0'
+```
+Now, we have to create directories in which to place our Rust library. When we built our library, we created a single `.so` file for each architecture we cross compiled for. We now need to create a subdirectory for each of these architectures inside a special directory called `jniLibs`. Android knows to look for them here. Once those directories are created, we can create a symlink to ensure that the Android Studio version of the libraries are always linking to the most recently built version.
+
+```
+cd ../android/Greetings/app/src/main
+mkdir jniLibs
+mkdir jniLibs/arm64
+mkdir jniLibs/armeabi
+mkdir jniLibs/x86
+
+ln -s <project_path>/greetings/cargo/target/aarch64-linux-android/release/libgreetings.so jniLibs/arm64/libgreetings.so
+ln -s <project_path>/greetings/cargo/target/armv7-linux-androideabi/release/libgreetings.so jniLibs/armeabi/libgreetings.so
+ln -s <project_path>/greetings/cargo/target/i686-linux-android/release/libgreetings.so jniLibs/x86/libgreetings.so
+```
+
+After we've linked out Rust library, we also need to add the `libjnidispatch.so` library for each architecture in the same directories. Right now, I am not sure where we get this from but I have links to pre-built versions of them here:
```

Comment:
In practice, I think, this might be linked-in via some CI process artifact. I think this is fine for the purposes of a basic tutorial though.

Reply:
JNA provides pre-built versions in platform-specific JAR archives here:

https://github.com/java-native-access/jna/tree/master/dist

And here:

https://github.com/java-native-access/jna/tree/master/lib/native

Which is more than a bit confusing! And [Android Development Environment](https://github.com/java-native-access/jna/blob/master/www/AndroidDevelopmentEnvironment.md) doesn't mention the platform-specific archives. I found them by following links in issue comments.

Reply:
This is certainly one bit of the post that I am going to make sure is better before I move this from `WIP`.